File: tests/e2e-sas-order-contract.test.mjs
/**
 * tests/e2e-sas-order-contract.test.mjs — INC-0924, RESUME-PROTOCOL RULE 30.
 *
 * ONE vector file, `tests/e2e-sas-order-vectors.json`, consumed by TWO
 * implementations:
 *
 *   - this file — the web half: the SAS window in `hooks/useE2e.ts`
 *     (`sasPendingRef` over `isSealedFrameType`), the teardown views in
 *     `hooks/phoneE2e.ts`, the copy in `lib/encryptedModeCopy.ts`, and the
 *     `PAIRING_TERMINATED` wiring in `hooks/usePhoneBridge.ts`;
 *   - dnkdialer-android/app/src/test/.../E2eSasOrderContractTest.kt — the
 *     phone half: the ACCEPT/prompt ORDER, `E2eFrameGate`, `E2eSasGate`.
 *
 * ## The assumption this suite exists to hold
 *
 * "The phone accepts BEFORE its user confirms, and nothing of the user's
 * crosses the wire until BOTH have." That is a claim about two codebases at
 * once, and neither one's own tests can see the other half. The specific way
 * it breaks is asymmetry: one side opens its window early, the other keeps
 * sending into it, and nobody is red.
 *
 * ## Why some checks read source
 *
 * `sealOutbound` and `openInbound` are closures inside a React hook and the
 * bridge's frame switch is a 2,000-line `case`. The FACT under test is one
 * line in each — that the SAS window is consulted at all, and that a
 * `PAIRING_TERMINATED` tells the e2e half the pair is gone. A suite that
 * re-implemented those lines would agree with itself and prove nothing, so
 * they are asserted against the real files, with the slice proved non-empty
 * first.
 *
 * Run: node tests/e2e-sas-order-contract.test.mjs
 */

import { createRequire } from 'node:module';
import { readFileSync } from 'node:fs';
import path from 'node:path';

import { viewAfterPairEnded, viewAfterPairEndedDuringSas } from '../hooks/phoneE2e.ts';
import { encryptionIndicator, E2E_ERRORS } from '../lib/encryptedModeCopy.ts';

const require = createRequire(import.meta.url);
const VECTORS = require('./e2e-sas-order-vectors.json');
const ROOT = path.resolve(import.meta.dirname, '..');

let passed = 0;
let failed = 0;
function check(name, ok, detail = '') {
  if (ok) { passed++; return true; }
  failed++;
  console.error(`  FAIL  ${name}${detail ? ` — ${detail}` : ''}`);
  return false;
}
const eq = (name, got, want) =>
  check(name, got === want, `got ${JSON.stringify(got)} want ${JSON.stringify(want)}`);

/** CRLF-safe (.gitattributes pins lib/hooks to LF; a checkout may not be). */
function stripComments(src) {
  return src
    .replace(/\r\n/g, '\n')
    .replace(/\/\*[\s\S]*?\*\//g, '')
    .replace(/^[ \t]*\/\/.*$/gm, '')
    .replace(/[ \t]\/\/.*$/gm, '');
}

const read = (...p) => stripComments(readFileSync(path.join(ROOT, ...p), 'utf8'));

const row = (id) => {
  const r = VECTORS.rows.find((x) => x.id === id);
  check(`row ${id} is present in the vector file`, Boolean(r));
  return r ?? {};
};

// ── 0. the file itself ──────────────────────────────────────────────────────
{
  eq('vector file version', VECTORS.version, 1);
  eq('the file still carries every ordering scenario', VECTORS.rows.length, 8);
  check('the ordering block demands ACCEPT before the prompt',
    VECTORS.ordering.acceptBeforePrompt === true);
  // The row whose expected value is "the product must NOT do this any more".
  // Deleting it would quietly delete the incident's memory.
  check('the old phone-first order is still recorded as forbidden',
    row('phone-first-confirm').forbidden === true);
  eq('and it records that the computer had no digits to compare',
    row('phone-first-confirm').browser.hasDigits, false);
}

// ── 1. the browser's SAS window is real, and it is consulted ────────────────
{
  const hook = read('hooks', 'useE2e.ts');

  const out = hook.slice(hook.indexOf('const sealOutbound = useCallback'));
  const outBody = out.slice(0, out.indexOf('\n  }, ['));
  check('the sealOutbound body actually sliced',
    outBody.includes('session.seal') && outBody.length > 200, `${outBody.length} chars`);
  check('sealOutbound refuses a sealed type while the SAS is pending',
    /sasPendingRef\.current && isSealedFrameType\(type\)/.test(outBody));
  check('and it REFUSES rather than passing the payload through',
    /throw new Error\(/.test(outBody.slice(outBody.indexOf('sasPendingRef'))));
  // The control: the guard is above the `!session` early return, so a pair
  // with no session still refuses instead of silently returning plaintext.
  check('the SAS guard sits ABOVE the no-session pass-through (control)',
    outBody.indexOf('sasPendingRef') < outBody.indexOf('if (!session'));

  const inb = hook.slice(hook.indexOf('const openInbound = useCallback'));
  const inBody = inb.slice(0, inb.indexOf('\n  }, ['));
  check('the openInbound body actually sliced',
    inBody.includes('session.open') && inBody.length > 200, `${inBody.length} chars`);
  check('openInbound DROPS a sealed type while the SAS is pending',
    /sasPendingRef\.current && isSealedFrameType\(type\) \) return \{ drop: true \}|sasPendingRef\.current && isSealedFrameType\(type\)\) return \{ drop: true \}/
      .test(inBody));

  // The window is keyed on the EFFECTIVE mode, not the sealing flag — the
  // SAS-MODE0 defect. Re-pinned here because this suite's whole subject is
  // when that window is open.
  check('sasPendingRef keys on the EFFECTIVE mode',
    /sasPendingRef\.current = decision\.effective === 'on'/.test(hook));
  check('a confirmed SAS is the only thing that opens the window on a live pair',
    /const confirmSas = useCallback[\s\S]{0,400}?sasPendingRef\.current = false;/.test(hook));

  // Rows that state a browser expectation about the window.
  for (const r of VECTORS.rows) {
    if (!r.browser || r.browser.sealedOutbound === undefined) continue;
    eq(`${r.id}: the browser refuses to seal inside the window`,
      r.browser.sealedOutbound, 'throw');
    check(`${r.id}: states the window is open`, r.browser.sasPending === true);
  }
}

// ── 2. a pair that ends mid-SAS says so ─────────────────────────────────────
{
  const base = {
    mode: 'on', effective: 'on', state: 'encrypted-verified',
    sas: { digits: '76386', confirmed: false },
    peer: { supports: true, kind: 'extension' },
  };

  const ended = viewAfterPairEndedDuringSas(base);
  eq('a teardown mid-SAS produces an error state', ended.state, 'error');
  eq('and names the SAS as the reason', ended.error, 'e2e-sas-unconfirmed');
  eq('the mode rides along so the badge can still say what was asked for',
    ended.mode, 'on');
  eq('the pair-scoped digits are gone', ended.sas.digits, null);
  eq('and the peer no longer supports (there is no peer)', ended.peer.supports, false);
  eq('peer.kind is a property of THIS browser and survives', ended.peer.kind, 'extension');

  // The control: the ordinary teardown must NOT invent this error, or the
  // assertion above would pass for every disconnect in the product.
  const plain = viewAfterPairEnded(base);
  eq('CONTROL: an ordinary teardown stays quiet', plain.state, 'unencrypted');
  eq('CONTROL: and sets no error', plain.error, undefined);

  // An error already on screen outranks it: a pair that refused for a named
  // reason must not have its story rewritten by a later teardown.
  const errored = { ...base, state: 'error', error: 'e2e-key-mismatch' };
  eq('an existing error survives a mid-SAS teardown',
    viewAfterPairEndedDuringSas(errored).error, 'e2e-key-mismatch');
  eq('and it is still an error', viewAfterPairEndedDuringSas(errored).state, 'error');

  // The two rows that arrive here.
  for (const id of ['phone-mismatch-abort', 'sas-timeout']) {
    const r = row(id);
    eq(`${id}: reaches the browser as PAIRING_TERMINATED`, r.browser.onFrame, 'PAIRING_TERMINATED');
    eq(`${id}: and the browser shows an error`, r.browser.state, 'error');
    eq(`${id}: named e2e-sas-unconfirmed`, r.browser.error, 'e2e-sas-unconfirmed');
    eq(`${id}: which the real view function produces`,
      viewAfterPairEndedDuringSas(base).error, r.browser.error);
  }
}

// ── 3. the wiring that makes rows 'phone-mismatch-abort' / 'sas-timeout' true ─
{
  const hook = read('hooks', 'useE2e.ts');
  const fn = hook.slice(hook.indexOf('const onPairEnded = useCallback'));
  const body = fn.slice(0, fn.indexOf('\n  }, ['));
  check('the onPairEnded body actually sliced',
    body.includes('sessionRef.current = null') && body.length > 200, `${body.length} chars`);
  check('onPairEnded reads the SAS window BEFORE clearing it',
    body.indexOf('sasPendingRef.current;') < body.indexOf('sasPendingRef.current = false'));
  check('and chooses the mid-SAS view when it was open',
    /viewAfterPairEndedDuringSas : viewAfterPairEnded/.test(body));

  const bridge = read('hooks', 'usePhoneBridge.ts');
  const tc = bridge.slice(bridge.indexOf("case 'PAIRING_TERMINATED': {"));
  const tcBody = tc.slice(0, tc.indexOf('\n      }'));
  check('the PAIRING_TERMINATED case actually sliced',
    tcBody.includes('setLobbyState') && tcBody.length > 200, `${tcBody.length} chars`);
  check('PAIRING_TERMINATED tells the e2e half the pair ended',
    /e2eRef\.current\.onPairEnded\(\)/.test(tcBody));
}

// ── 4. the copy ─────────────────────────────────────────────────────────────
{
  check('the new error code is in the frozen list for the node harness',
    E2E_ERRORS.includes('e2e-sas-unconfirmed'));
  const ind = encryptionIndicator({
    mode: 'on', effective: 'on', state: 'error', error: 'e2e-sas-unconfirmed',
    sas: { digits: null, confirmed: false }, peer: { supports: false, kind: null },
  });
  check('it raises a banner', ind.banner === true);
  eq('with its own label, not the generic refusal', ind.label, 'Code not confirmed');
  check('the detail says the pairing ended before both codes were confirmed',
    /before both codes were confirmed/.test(ind.detail), ind.detail);
  // Retrying is what an attacker needs; the copy must not offer it as the
  // obvious next tap. The phone's own refusal string follows the same rule.
  check('and it does not invite a retry', !/try again/i.test(ind.detail), ind.detail);
  // CONTROL: a different error must not produce this copy, or the assertions
  // above would be satisfied by the default branch.
  const other = encryptionIndicator({
    mode: 'on', effective: 'on', state: 'error', error: 'e2e-setup-failed',
    sas: { digits: null, confirmed: false }, peer: { supports: false, kind: null },
  });
  check('CONTROL: the generic refusal still reads as itself',
    other.label !== 'Code not confirmed', other.label);
}

// ── 5. the kill-switch row: recorded, not endorsed ──────────────────────────
{
  const r = row('killswitch-off-phone-on');
  check('the kill-switch row is marked documented-not-endorsed',
    r.documentedNotEndorsed === true);
  check('and it states that the phone still forces a verified pair',
    r.phone.effectiveMode === 'ENCRYPTED_VERIFIED');
  check('its note points at the open question rather than resolving it',
    /Security/.test(r.why), r.why);
}

console.log(`\n${failed === 0 ? 'PASS' : 'FAIL'}  ${passed} passed, ${failed} failed`);
process.exit(failed === 0 ? 0 : 1);

File: dnkdialer-android/app/src/test/java/com/dnkdialer/companion/E2eSasOrderContractTest.kt
package com.dnkdialer.companion

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * INC-0924 — the phone half of `tests/e2e-sas-order-vectors.json`
 * (RESUME-PROTOCOL v3.0 RULE 30).
 *
 * The web half is `tests/e2e-sas-order-contract.test.mjs`, reading the same
 * file. Two surfaces implement one ordering, and the only thing that catches
 * them disagreeing is a fixture neither of them authored alone.
 *
 * ## Three different instruments, on purpose
 *
 *  1. **The ORDER** is read out of `PhoneService.kt` itself. The accept path
 *     needs a relay socket, the Keystore and a bound Activity, so it has no
 *     JVM seam — and the defect was not a wrong value, it was two correct
 *     statements in the wrong order. A source assertion is the only instrument
 *     that can see that, and it goes red the moment they are swapped back.
 *  2. **The WINDOW** is asserted against a REAL [E2eFrameGate], constructed
 *     the way [PhoneService] constructs it. Not a re-implementation of the
 *     rule: the product object, told it is mid-SAS, asked what it does with a
 *     §13.7 frame.
 *  3. **The VERDICTS** go through the real [E2eSasGate.mayProceed].
 *
 * Run: `gradlew.bat testDebugUnitTest --tests '*E2eSasOrderContract*'`
 */
class E2eSasOrderContractTest {

    /** Unit-test cwd is `dnkdialer-android/app`, so `../..` is the repo root. */
    private val file = File("../../tests/e2e-sas-order-vectors.json")

    private val service = File("src/main/java/com/dnkdialer/companion/PhoneService.kt")

    private fun root(): JsonObject {
        assertTrue("the shared vector file is missing at " + file.absolutePath, file.exists())
        val root = JsonParser.parseString(file.readText()).asJsonObject
        assertEquals("vector file version", 1, root.get("version").asInt)
        return root
    }

    private fun rows(): List<JsonObject> {
        val rows = root().getAsJsonArray("rows").map { it.asJsonObject }
        // A vectors test whose file lost its rows passes vacuously.
        assertEquals("the ordering scenarios", 8, rows.size)
        return rows
    }

    private fun row(id: String): JsonObject =
        rows().firstOrNull { it.get("id").asString == id }
            ?: throw AssertionError("the vector file no longer carries row $id")

    private fun phone(row: JsonObject): JsonObject? =
        row.getAsJsonObject("phone")

    // ------------------------------------------------------------ (1) ORDER

    /**
     * The fix itself: `ACCEPT_PAIRING` leaves BEFORE the user is prompted.
     *
     * Sliced from the accept function only, so a match anywhere else in this
     * 4,000-line file cannot satisfy it.
     */
    @Test
    fun the_accept_is_sent_before_the_user_is_prompted() {
        val ordering = root().getAsJsonObject("ordering")
        assertTrue(
            "the vector file must demand the ACCEPT-first order",
            ordering.get("acceptBeforePrompt").asBoolean
        )
        assertTrue("PhoneService.kt not found at " + service.absolutePath, service.exists())
        // CRLF-safe: a fresh checkout on Windows is CRLF and every regex below
        // would otherwise be reading prose it thinks is code.
        val src = service.readText().replace("\r\n", "\n")

        val fnStart = src.indexOf("private fun completeEncryptedAccept(")
        assertTrue("completeEncryptedAccept has been renamed or removed", fnStart >= 0)
        val fnEnd = src.indexOf("\n    private fun broadcastE2eRefusal", fnStart)
        assertTrue("could not find the end of the accept path", fnEnd > fnStart)
        val body = src.substring(fnStart, fnEnd)
        // A slice that silently came back empty passes every assertion below
        // for free, so prove it is the real body first.
        assertTrue("the accept body did not slice (got ${body.length} chars)", body.length > 2000)

        val accept = body.indexOf(ordering.get("acceptCall").asString)
        val prompt = body.indexOf(ordering.get("promptCall").asString)
        assertTrue("the ACCEPT send is gone from the accept path", accept >= 0)
        assertTrue("the SAS prompt is gone from the accept path", prompt >= 0)
        assertTrue(
            "INC-0924: ACCEPT_PAIRING must be sent BEFORE E2eSasGate.await — the computer " +
                "derives its digits from that ACCEPT, so prompting first asks the user to " +
                "compare a code against a blank screen (accept@$accept prompt@$prompt)",
            accept < prompt
        )

        // And the refusal after an ACCEPT must end the ACTIVE pair, not answer
        // a request the relay has already resolved.
        val refusal = row("phone-mismatch-abort").getAsJsonObject("phone")
            .get("teardownFrame").asString
        assertTrue(
            "a SAS refusal after the ACCEPT must tear the pair down with $refusal",
            body.contains("leaveActivePair(\"SAS not confirmed\")")
        )
        assertTrue(
            "$refusal must be the frame leaveActivePair sends",
            src.contains("sendResponse(\"$refusal\"")
        )
    }

    /**
     * The window has to be ARMED before the ACCEPT leaves, not after: the peer
     * may answer the instant it lands, and a gate armed one statement later
     * would have let that first frame through.
     */
    @Test
    fun the_frame_window_is_armed_before_the_accept_leaves() {
        val src = service.readText().replace("\r\n", "\n")
        val fnStart = src.indexOf("private fun completeEncryptedAccept(")
        val fnEnd = src.indexOf("\n    private fun broadcastE2eRefusal", fnStart)
        val body = src.substring(fnStart, fnEnd)
        assertTrue("the accept body did not slice", body.length > 2000)

        val arm = body.indexOf("e2eSasPending = prepared.modeOn")
        val accept = body.indexOf("sendPairingDecision(\"ACCEPT_PAIRING\"")
        assertTrue("the SAS window is never armed", arm >= 0)
        assertTrue(
            "e2eSasPending must be set BEFORE the ACCEPT leaves (arm@$arm accept@$accept)",
            arm in 0 until accept
        )
        assertTrue(
            "the gate must read the window",
            src.contains("E2eFrameGate({ e2eSession }, { e2eLatchedOn }, { e2eSasPending })")
        )
    }

    // ----------------------------------------------------------- (2) WINDOW

    /**
     * The gate, told it is mid-SAS, over every row that states an expectation.
     *
     * `PHONE_NOTIFICATION` is the §13.7 type the phone actually emits and
     * `SEND_SMS` the one it actually receives, so the two directions are
     * exercised with frames the product really carries rather than a
     * hand-picked easy case.
     */
    @Test
    fun the_frame_gate_matches_every_row_that_states_one() {
        var checked = 0
        for (r in rows()) {
            val id = r.get("id").asString
            val p = phone(r) ?: continue
            val pending = p.get("sasPending")?.asBoolean ?: false
            // The gate as PhoneService builds it: a live session, latched, and
            // the window's flag. A null session would make every answer a drop
            // for the wrong reason, so the sealed path has to be reachable.
            val gate = E2eFrameGate(
                sessionProvider = { null },
                latchedProvider = { true },
                sasPendingProvider = { pending },
            )

            p.get("sealedOutbound")?.let { want ->
                checked++
                val out = gate.outbound("PHONE_NOTIFICATION", "{\"title\":\"x\"}")
                when (want.asString) {
                    "drop" -> {
                        assertEquals("$id: a sealed frame must be DROPPED, never downgraded", null, out)
                        assertTrue(
                            "$id: the drop must name the SAS window, not the latch",
                            gate.lastDropReason?.contains("sas pending") == true
                        )
                    }
                    // "send" cannot be asserted against a null session (there
                    // is nothing to seal with); what it asserts is that the
                    // WINDOW is not what stops it. The reason proves which
                    // arm refused.
                    "send" -> assertFalse(
                        "$id: the SAS window must be open after a MATCH",
                        gate.lastDropReason?.contains("sas pending") == true
                    )
                    else -> throw AssertionError("$id: unknown sealedOutbound ${want.asString}")
                }
            }

            p.get("sealedInbound")?.let { want ->
                checked++
                val gateIn = E2eFrameGate(
                    sessionProvider = { null },
                    latchedProvider = { true },
                    sasPendingProvider = { pending },
                )
                val res = gateIn.inbound("SEND_SMS", "{\"to\":\"x\"}")
                if (want.asString == "drop") {
                    assertTrue("$id: inbound must be dropped", res is E2eFrameGate.Inbound.Drop)
                    assertTrue(
                        "$id: dropped for the SAS window",
                        (res as E2eFrameGate.Inbound.Drop).reason.contains("sas pending")
                    )
                } else {
                    // Not the window's doing, whatever else happens to it.
                    val reason = (res as? E2eFrameGate.Inbound.Drop)?.reason ?: ""
                    assertFalse("$id: the window must be open", reason.contains("sas pending"))
                }
            }
        }
        assertTrue("no row stated a frame expectation — the file lost its rows", checked >= 8)
    }

    /**
     * A frame type §13.7 leaves in the clear is untouched by the window. The
     * window withholds USER DATA; it is not a general mute, and muting
     * `GET_MESSAGES` would move tier enforcement off the relay.
     */
    @Test
    fun the_window_never_touches_a_plaintext_type() {
        val gate = E2eFrameGate({ null }, { true }, { true })
        assertEquals("{}", gate.outbound("GET_MESSAGES", "{}"))
        val res = gate.inbound("GET_MESSAGES", "{}")
        assertTrue(res is E2eFrameGate.Inbound.Deliver)
        assertEquals(0L, gate.droppedOutbound)
    }

    // --------------------------------------------------------- (3) VERDICTS

    @Test
    fun every_verdict_in_the_file_matches_may_proceed() {
        var checked = 0
        for (r in rows()) {
            val id = r.get("id").asString
            val p = phone(r) ?: continue
            val v = p.get("verdict") ?: continue
            if (v.isJsonNull) continue
            val want = p.get("mayProceed")?.asBoolean ?: continue
            checked++
            assertEquals(
                "$id: ${E2eSasGate.Verdict.valueOf(v.asString)}",
                want,
                E2eSasGate.mayProceed(E2eSasGate.Verdict.valueOf(v.asString))
            )
        }
        assertTrue("no verdict rows survived in the file", checked >= 4)
        // MATCHED and NOT_REQUIRED are the ONLY two. Silence, a cancel and a
        // "doesn't match" are one outcome downstream by design.
        for (verdict in E2eSasGate.Verdict.values()) {
            val expected = verdict == E2eSasGate.Verdict.MATCHED ||
                verdict == E2eSasGate.Verdict.NOT_REQUIRED
            assertEquals("$verdict", expected, E2eSasGate.mayProceed(verdict))
        }
    }

    /**
     * The forbidden row. It has no value to compare — what it pins is that the
     * product no longer contains the old shape, which the ordering test above
     * asserts positively. This case exists so that DELETING that row from the
     * file is a failure rather than a quiet loss of the incident's memory.
     */
    @Test
    fun the_old_phone_first_ordering_is_still_recorded_as_forbidden() {
        val old = row("phone-first-confirm")
        assertTrue("the phone-first row must stay marked forbidden", old.get("forbidden").asBoolean)
        val p = old.getAsJsonObject("phone")
        assertFalse("the old order sent no ACCEPT before prompting", p.get("acceptSent").asBoolean)
        assertTrue("the old order prompted", p.get("sasPrompted").asBoolean)
        assertFalse(
            "and the computer had no digits to compare against",
            old.getAsJsonObject("browser").get("hasDigits").asBoolean
        )
    }

    /**
     * The kill-switch row is recorded, NOT endorsed. Whether a phone with a
     * local ON should still force a SAS-blocking pair while
     * `E2E_PAIRING_ENABLED=0` is a Security question, and this lane's job is
     * to make the current answer visible rather than to change it.
     */
    @Test
    fun the_kill_switch_row_is_documented_not_endorsed() {
        val r = row("killswitch-off-phone-on")
        assertTrue(r.get("documentedNotEndorsed").asBoolean)
        val p = r.getAsJsonObject("phone")
        // The behaviour itself, through the real decision table: the phone's
        // own ON is enough, whatever the relay's switch says.
        assertEquals(
            E2eSettings.EffectiveMode.valueOf(p.get("effectiveMode").asString),
            E2eSettings.effectiveMode(
                localEnabled = true,
                peer = E2eSettings.PeerAdvertisement.OFF,
            )
        )
        assertEquals(
            p.get("requiresSas").asBoolean,
            E2eSettings.requiresSas(E2eSettings.EffectiveMode.ENCRYPTED_VERIFIED)
        )
    }

    // ------------------------------------------- (4) THE STATUS THE PHONE SHOWS

    /**
     * INC-0924 Security F-1 (C1). What the phone TELLS ITS OWN USER while the
     * SAS dialog is unanswered.
     *
     * `e2eVerified` is the key-pin verdict alone. Against a TOFU-pinned peer
     * it is already `true` when `broadcastE2eState()` runs one line above
     * `sendPairingDecision("ACCEPT_PAIRING", ...)` — i.e. at the TOP of the
     * window — so the un-amended predicate badged the pair "Encrypted and
     * verified" at the exact moment it was asking the user to verify it. That
     * is a tap-through prime, and a tap-through is the only thing that defeats
     * this SAS.
     *
     * Two instruments, because [E2eStatusCopy.stateOf] is unchanged and a value
     * test alone would pass over the old one-arg call:
     *  - the VALUE, through the real `stateOf`, over every row that states a
     *    `statusState` (`accept-then-both-confirm` is the control: a predicate
     *    that answered UNVERIFIED unconditionally passes the window row and
     *    fails that one);
     *  - the PREDICATE, read out of `PhoneService.currentE2eState` itself,
     *    which is the only instrument that sees `&& !e2eSasPending` go missing.
     */
    @Test
    fun the_phone_reports_unverified_for_the_whole_sas_window() {
        var checked = 0
        for (r in rows()) {
            val id = r.get("id").asString
            val p = phone(r) ?: continue
            val want = p.get("statusState")?.asString ?: continue
            val pinned = p.get("keyPinVerified").asBoolean
            val pending = p.get("sasPending").asBoolean
            checked++
            assertEquals(
                "$id: keyPinVerified=$pinned sasPending=$pending",
                E2eStatusCopy.State.valueOf(want),
                E2eStatusCopy.stateOf(encrypted = true, verified = pinned && !pending),
            )
        }
        assertEquals("both status rows must survive in the file", 2, checked)

        val src = service.readText().replace("\r\n", "\n")
        val fnStart = src.indexOf("fun currentE2eState()")
        assertTrue("currentE2eState has been renamed or removed", fnStart >= 0)
        val fnEnd = src.indexOf("private fun broadcastE2eState", fnStart)
        assertTrue("could not find the end of currentE2eState", fnEnd > fnStart)
        val body = src.substring(fnStart, fnEnd)
        assertTrue("currentE2eState did not slice (got ${body.length} chars)", body.length > 120)
        assertTrue(
            "Security F-1: currentE2eState must report UNVERIFIED while the phone's own " +
                "SAS dialog is unanswered — verified = e2eVerified && !e2eSasPending",
            body.contains("verified = e2eVerified && !e2eSasPending"),
        )
    }

    /**
     * C2 (Security F-3). The refusal counters are the ONLY forensic trace of
     * this window — a field MITM attempt leaves nothing else behind. There is
     * no allowlist to add them to: `DiagLog.counter` is an open registry and
     * every name it is given is persisted to `counters.json` and shipped
     * verbatim in the export zip. So what needs pinning is that the two call
     * sites still exist, spelled the way a field zip will be grepped for.
     */
    @Test
    fun the_window_refusals_leave_a_counter_behind() {
        val src = service.readText().replace("\r\n", "\n")
        for (name in listOf("e2e.sas.refused-after-accept", "e2e.sas.malformed")) {
            assertTrue(
                "the only forensic trace of a refused SAS is DiagLog.counter(\"$name\")",
                src.contains("DiagLog.counter(\"$name\")"),
            )
        }
    }
}
